bokehjs/test/models/annotations/legend_item.coffee

```diff
@@ -0,0 +1,150 @@
+      LegendItem.prototype._check_data_sources_on_renderers.restore()
+
+    it "should log an error if _check_field_label_on_data_source is false", ->
+      sinon.stub(LegendItem.prototype, '_check_field_label_on_data_source').returns(false)
+      legend_item = new LegendItem()
+      sinon.assert.calledOnce(logger.error)
+      LegendItem.prototype._check_field_label_on_data_source.restore()
+
+  describe "_check_data_sources_on_renderers", ->
+
+    it "should return false if field label and different data sources", ->
+      gr_1 = new GlyphRenderer({ data_source: new ColumnDataSource() })
+      gr_2 = new GlyphRenderer({ data_source: new ColumnDataSource() })
+      legend_item = new LegendItem({
+        label: { field: 'label' },
+        renderers: [ gr_1, gr_2 ]
+      })
+      expect(legend_item._check_data_sources_on_renderers()).to.be.false
+
+    it "should return false if field label and no renderers", ->
+      legend_item = new LegendItem({
+        label: { field: 'label' },
+        renderers: [ ]
+      })
+      expect(legend_item._check_data_sources_on_renderers()).to.be.false
+
+    it "should return true if value label and different data sources", ->
+      gr_1 = new GlyphRenderer({ data_source: new ColumnDataSource() })
+      gr_2 = new GlyphRenderer({ data_source: new ColumnDataSource() })
+      legend_item = new LegendItem({
+        label: { value: 'label' },
+        renderers: [ gr_1, gr_2 ]
+      })
+      expect(legend_item._check_data_sources_on_renderers()).to.be.true
+
+  describe "_check_field_label_on_data_source", ->
+
+    it "should return false if field label and label not in data source", ->
+      gr_1 = new GlyphRenderer({
+        data_source: new ColumnDataSource({data: {foo: [1]}})
+      })
+      legend_item = new LegendItem({
+        label: { field: 'label' },
+        renderers: [ gr_1 ]
+      })
+      expect(legend_item._check_field_label_on_data_source()).to.be.false
+
+    it "should return false if field label and no renderers", ->
+      legend_item = new LegendItem({
+        label: { field: 'label' },
+        renderers: [ ]
+      })
+      expect(legend_item._check_field_label_on_data_source()).to.be.false
+
+    it "should return true if field label and label in data source", ->
+      gr_1 = new GlyphRenderer({
+        data_source: new ColumnDataSource({data: {label: [1]}})
+      })
+      legend_item = new LegendItem({
+        label: { field: 'label' },
+        renderers: [ gr_1 ]
+      })
+      expect(legend_item._check_field_label_on_data_source()).to.be.true
+
+  describe "get_field_from_label_prop", ->
+
+    it "should return undefined if label property is null", ->
+      legend_item = new LegendItem({'label': null})
+      field = legend_item.get_field_from_label_prop()
+      expect(field).to.be.undefined
+
+    it "should return undefined if label property is value", ->
+      legend_item = new LegendItem({'label': {'value': 'milk'}})
```

Comment:
This code was copied, so no blame here, but I would like to point out that it's preferred it use name literals instead of strings when referring to members of an interface, i.e. `new LegendItem({label: {value: 'milk'}})`. This states the intent better, because, after all, `label` and `value` are members of some interfaces and we aren't building an arbitrary dictionary here. Currently implicit, but when we will use TypeScript, it will be defined somewhere. Actually, it is already under `api/typings`. This is similar to what we do in Python when constructing keyword arguments with `dict(name=value)`, instead of `{"name": value}`.